test/integration/operator/garden/reference/reference_test.go

```diff
@@ -0,0 +1,243 @@
+							AuditConfig: &gardencorev1beta1.AuditConfig{
+								AuditPolicy: &gardencorev1beta1.AuditPolicy{
+									ConfigMapRef: &corev1.ObjectReference{
+										Name: configMap2.Name,
+									},
+								},
+							},
+							AuditWebhook: &operatorv1alpha1.AuditWebhook{
+								KubeconfigSecretName: secret7.Name,
+							},
+						},
+					},
+					Maintenance: operatorv1alpha1.Maintenance{
+						TimeWindow: gardencorev1beta1.MaintenanceTimeWindow{
+							Begin: "220000+0100",
+							End:   "230000+0100",
+						},
+					},
+					Networking: operatorv1alpha1.Networking{
+						Services: "100.64.0.0/13",
+					},
+				},
+			},
+		}
+	})
+
+	JustBeforeEach(func() {
+		By("Create Garden")
+		Expect(testClient.Create(ctx, garden)).To(Succeed())
+		log.Info("Created Garden for test", "garden", client.ObjectKeyFromObject(garden))
+
+		DeferCleanup(func() {
+			By("Delete Garden")
+			Expect(client.IgnoreNotFound(testClient.Delete(ctx, garden))).To(Succeed())
+
+			By("Ensure Garden is gone")
+			Eventually(func() error {
+				return mgrClient.Get(ctx, client.ObjectKeyFromObject(garden), garden)
+			}).Should(BeNotFoundError())
+		})
+	})
+
+	Context("no references", func() {
+		BeforeEach(func() {
+			garden.Spec.VirtualCluster.ETCD = nil
+			garden.Spec.VirtualCluster.Kubernetes.KubeAPIServer = nil
+			garden.Spec.VirtualCluster.Gardener.APIServer = nil
+		})
+
+		It("should not add the finalizer to the garden", func() {
+			Consistently(func(g Gomega) []string {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(garden), garden)).To(Succeed())
+				return garden.Finalizers
+			}).ShouldNot(ContainElement("gardener.cloud/reference-protection"))
+		})
+	})
+
+	Context("w/ references", func() {
+		JustBeforeEach(func() {
+			Eventually(func(g Gomega) []string {
+				g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(garden), garden)).To(Succeed())
+				return garden.Finalizers
+			}).Should(ContainElement("gardener.cloud/reference-protection"))
+		})
+
+		It("should add finalizers to the referenced secrets and configmaps", func() {
+			for _, obj := range []client.Object{secret1, secret2, secret3, secret4, secret5, secret6, secret7, configMap1, configMap2} {
+				Eventually(func(g Gomega) []string {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(obj), obj)).To(Succeed())
+					return obj.GetFinalizers()
+				}).Should(ContainElement("gardener.cloud/reference-protection"), obj.GetName()+" should have the finalizer")
+			}
+		})
+
+		It("should remove finalizers from the referenced secrets and configmaps", func() {
+			patch := client.MergeFrom(garden.DeepCopy())
+			garden.Spec.VirtualCluster.ETCD = nil
+			garden.Spec.VirtualCluster.Kubernetes.KubeAPIServer = nil
+			garden.Spec.VirtualCluster.Gardener.APIServer = nil
+			Expect(testClient.Patch(ctx, garden, patch)).To(Succeed())
+
+			for _, obj := range []client.Object{secret1, secret2, secret3, secret4, secret5, secret6, secret7, configMap1, configMap2} {
```

Comment:
reference protection annotation should be removed from garden also.
```suggestion
			for _, obj := range []client.Object{secret1, secret2, secret3, secret4, secret5, secret6, secret7, configMap1, configMap2, garden} {
```